### ISSUE TITLE
Tracer takes an ErrorHandler option

### ISSUE THREAD

**Issue by tedsuo** [lightstep/lightstep-tracer-go#91]
Currently the lightstep tracer swallows it's errors, as there is generally no actionable response a user can take and the OT API does not support returned errors. However, for reporting purposes a centralized ErrorHandler could be added to the Tracer as an Option:

```go
type ErrorHandler func(error)
```

Error categories must be identified so that reporters can make metrics for them:
- configuration errors: the tracer will not start due to incorrect setup
- reporting errors: the tracer cannot connect to the configured collector

Question: currently the stand-alone functions `CloseTracer`, `FlushLightstepTracer` and `GetAccessToken` return errors directly. Likewise, the Tracer interface currently returns an error on Close, to conform to `io.Closer`. For errors returned by these functions which could also be reported to the ErrorHandler callback, do we report them in both places?

**Comment by tedsuo** [comment]
https://github.com/lightstep/lightstep-tracer-go/issues/89#issuecomment-314618694 contains a more detailed spec.

**Comment by ChimeraCoder** [comment]
I'm a strong proponent of synchronous error handling - it's easy to turn synchronous code in Go to asynchronous code if needed, and much clunkier to do the reverse.

So I'd be in favor of having those functions return errors directly, because that provides callers with a path to handle all errors in the foreground, if they prefer to.


If that's not possible, I'd actually suggest channel-based delivery of errors instead of a callback; it's more convenient to be able to combine error handling with the other concurrency logic that come for free with channels than to handle it with callbacks.

**Comment by tedsuo** [comment]
Hi @ChimeraCoder, thanks for your feedback. We're now actively working on a PR for error handling (https://github.com/lightstep/lightstep-tracer-go/pull/101). We're going to go with callbacks, but I wanted to give you some reasoning.

## What we expect users to do with errors
We expect users, for the most part, to use error handling to report errors through their logging and metrics solutions of choice. We are considering adding other diagnostic callbacks as well, but right now we are starting with errors.

## On returning errors from Flush and Close
Normally I agree that synchronous errors (and synchronous methods in general) are the correct way to write go code. However, for two reasons we would prefer not to in this case.

The first reason is that most errors cannot be handled synchronously, because they occur while the tracer is in flight and not in response to a method call. That means there must be an asynchronous error handler of some sort. For errors on Flush and Close, we have the choice of double reporting (both returning directly and pushing to the error handler), split reporting (the error handler does not receive those errors), or only reporting to the error handler.

The second reason is we don't believe there is a good reason to handle errors returned by flush or close as a control flow mechanism. This is because the only reason that have been suggested for this basically amounts to adding external retry logic on top of the internal retry logic... we think that is a bad pattern, and would prefer any issues with retry be pushed down into the tracer. Instead, we would prefer users pass in a Context if they would like to control how long to wait for Flush or Close. If a Context is passed in, we will retry until the Context is done, so that the caller can coordinate shutdown, flushing, etc. If no Context is passed in, the tracer will eventually timeout based on it's own judgement.

The current convenience methods that wrap the tracer interface will continue to return an error for backwards compatibility, but it will only alert you to a typecasting failure (AKA you passed it something other than a LightStep tracer). We will add new convenience methods which don't return an error, and deprecate the current ones. The typecasting error is pretty marginal, as the type signature we are checking against is `opentracing.Tracer`, and we don't think it is a realistic production scenario where users are calling lightstep methods but actually using a different tracer. A dreaded `println` will do in that case as the call is a no-op and, again, extremely unlikely.

## On having a channel-based interface
Go's concurrency model is great, but IMHO channels are a bit of a nightmare as interfaces, as almost any choice you make is "bad" from someone's perspective. For example, do we buffer, block, or drop errors in the face of backpressure? I don't think we could pick one option and make everyone happy. But as a convenience, we can add an error handler to the lightstep package that has what we would suggest as the most commonly appropriate channel setup:

```go
// NewErrorChannel returns an OnError callback handler, and a channel that 
// produces the errors. When the channel buffer is full, subsequent errors will 
// be dropped.
NewErrorChannel(bufferSize int) (ErrorHandler, <-chan error)
```

This way, we leave the door open for users to implement other options if they need to, or just use the callback if all they are doing is reporting the error.

Hopefully this addresses your concerns, and please let us know if you have any questions or suggestions.

**Comment by tedsuo** [comment]
This issue was resolved in v0.15 by adding an Events framework for recording errors and diagnostics.

https://github.com/lightstep/lightstep-tracer-go#event-handling-observing-the-lightstep-tracer

Cheers,
-Ted
